### ISSUE TITLE
Production?

### ISSUE THREAD

**Issue by stokesbga** [sane/sails-hook-babel#20]
Does anyone have an elegant solution for building a sails app with babel for production?

**Comment by abecks** [comment]
Hey @stokesbga, were you ever able to find a solution?

**Comment by vikr01** [comment]
@abecks Don't use sails-hook-babel. It just uses babel-register, which caches your files and should not be used in production. It also won't let you use babel on your config files. It also comes with presets and plugins which may not help you.

I've started using a [yarn-workspaces](https://yarnpkg.com/lang/en/docs/workspaces/) ([lerna is another good option](https://github.com/lerna/lerna) if you're using npm) monorepo with [my modified version of sails](https://github.com/vikr01/sails/tree/feature/babel-support) in it that allows me to use babel properly. You can also do `yarn add -D https://github.com/vikr01/sails.git#feature/babel-support` (or `npm i -D https://github.com/vikr01/sails.git#feature/babel-support`) to install directly from this repo (I recommend the monorepo approach though, so you can maintain it yourself).

This is similar to what the structure of my sails project now looks like:
```bash
.
├── .babelrc
├── .env
├── .eslintignore
├── .eslintrc
├── .sailsrc
├── assets
├── package.json
├── src
│   ├── api
│   ├── app.js
│   ├── config
│   └── views
└── test
```

You'll need to add these scripts in your `package.json`:
```json
{
  "scripts": {
    "build": "babel src --out-dir dist --source-maps --copy-files",
    "dev": "node -r babel-register src/app.js",
    "start": "node dist/app.js"
  }
}
```
To start your project in development, you'd run `yarn dev` or `npm run dev`. To start your project in production you'd run `yarn build && yarn start` or `npm run build && npm start`.

Dependencies you'll need: `babel-register`, `babel-cli`.

My `app.js` looks like:
```javascript
import 'dotenv/config';
import path from 'path';
import rc from 'sails/accessible/rc';
import sails from 'sails';

const config = rc('sails');
config.appPath = __dirname;
Object.assign(config.paths, {
  public: path.join(__dirname, '..', 'assets'),
});
sails.lift(config);
```

I'm using `babel-preset-env` and quite a few other plugins in my .babelrc.

I submitted these changes in a [PR to sails](https://github.com/balderdashy/sails/pull/4462) a while ago, but seems it was missed.

**Comment by vikr01** [comment]
@abecks Don't use `sails-hook-babel`. It just uses [babel-register](https://github.com/babel/babel/tree/6.x/packages/babel-register), which caches your files and should not be used in production. It also won't let you use babel on your config files. It also comes with presets and plugins which may not help you. You lose too much flexibility when using `sails-hook-babel`.

I've started using a [yarn-workspaces](https://yarnpkg.com/lang/en/docs/workspaces/) ([lerna is another good option](https://github.com/lerna/lerna) if you're using npm) monorepo with [my modified version of sails](https://github.com/vikr01/sails/tree/feature/babel-support), [sails-hook-orm](https://github.com/vikr01/sails-hook-orm), and [sails-disk](https://github.com/vikr01/sails-disk) in it that allows me to use babel properly.

You can also do
`yarn add -D https://github.com/vikr01/sails.git#feature/babel-support https://github.com/vikr01/sails-disk.git#fix/tmp-path https://github.com/vikr01/sails-hook-orm.git#feature/babel-support`

or
`npm i -D https://github.com/vikr01/sails.git#feature/babel-support https://github.com/vikr01/sails-disk.git#fix/tmp-path https://github.com/vikr01/sails-hook-orm.git#feature/babel-support`

to install directly from my repos (I recommend the monorepo approach though, so you can maintainthem yourself).

This is similar to what the structure of my sails project now looks like:
```bash
.
├── .babelrc
├── .env
├── .eslintignore
├── .eslintrc
├── .sailsrc
├── assets
├── package.json
├── src
│   ├── api
│   ├── app.js
│   ├── config
│   └── views
└── test
```

How my project looks like as a monorepo with sails and my project in it:
```bash
.
├── package.json
├── packages
│   ├── my-sails-project
│   ├── sails
│   ├── sails-disk
│   └── sails-hook-orm
└── yarn.lock
```

You'll need to add these scripts in your `package.json`:
```json
{
  "scripts": {
    "build": "babel src --out-dir dist --source-maps --copy-files",
    "dev": "node -r babel-register src/app.js",
    "start": "node dist/app.js"
  }
}
```
To start your project in development, you'd run `yarn dev` or `npm run dev`. To start your project in production you'd run `yarn build && yarn start` or `npm run build && npm start`.

devDependencies you'll need: `babel-register`, `babel-cli`.

My `app.js` looks like:
```javascript
import 'dotenv/config';
import path from 'path';
import rc from 'sails/accessible/rc';
import sails from 'sails';

const config = rc('sails');
config.appPath = __dirname;
config.paths = Object.assign(config.paths || {}, {
  public: path.join(__dirname, '..', 'assets'),
});
sails.lift(config);
```

I'm using `babel-preset-env` and quite a few other plugins in my .babelrc.

I submitted these changes in a [PR to sails](https://github.com/balderdashy/sails/pull/4462) a while ago, but seems it was missed.

**Comment by stokesbga** [comment]
Yep, pretty much what I was going to say. Use babel-cli to generate a build folder and deploy that.

I have a package script that creates build folder, ignores some shit I only use in dev, then deploys that to production branch in git. Production branch has a hook that deploys.

````"build": "npm run clean && babel -d ./build . -s --ignore build,node_modules --copy-files && rm -r build/keys && rm -r build/config/babel.js && rm -r build/config/babel.js.map",```
